feat: optimize product indexing to avoid unnecessary table refresh

vienthuong opened the pull request:
Introduces diff-based updates for product category trees, stream mappings, search keywords, cheapest_price, and variant listing config so that only changed rows are inserted/deleted instead of tearing down whole tables.  Most of the time, these fields don't change, so always wiping them out and inserting a new one is a waste of resources.

This example result is for indexing with 100 products; it's just for reference; the result may vary. The test assumed that there were no changes in products and that the refresh index was run without any "real" update. Previously, there were many insert/delete statements (refresh tables) being executed even though the table was not changed. With the new implementation, there are no tables refresh if nothing has changed

Before:

<img width="495" height="159" alt="image" src="https://github.com/user-attachments/assets/5a05d83f-4665-452c-9762-c448d1da1af2" />


After:

<img width="515" height="170" alt="image" src="https://github.com/user-attachments/assets/054c81b1-78c2-46e3-b6a8-36c4e68b81f0" />

vienthuong committed:
feat: optimize product indexing to avoid unnecessary table refresh

vienthuong committed:
feat: optimize product indexing to avoid unnecessary table refresh

vienthuong committed:
feat: skip update cheapest_price if unchanged

vienthuong committed:
feat: skip update cheapest_price if unchanged

vienthuong committed:
feat: skip update variant listing if unchanged

vienthuong committed:
feat: skip update variant listing if unchanged

keulinho commented on src/Core/Content/Product/DataAbstractionLayer/CheapestPrice/CheapestPriceContainer.php:
```diff
@@ -28,11 +29,15 @@ class CheapestPriceContainer extends Struct
      */
     private ?array $ruleIds = null;
 
+    private string $hashed = '';
```
let's not provide a default value, so it is clear this should be filled in __construct

vienthuong committed:
fix(core): delete obsolete category tree rows in version contexts

The diff based delete matched `category_version_id` against the context version,
while the inserts always write the live category version. In a non live version
context the two never match, so obsolete `product_category_tree` rows were never
removed.

vienthuong committed:
fix(core): remove stream mappings that are no longer evaluated

`updateProducts()` only deleted the diff of every stream it evaluated, so mappings
of streams that have been invalidated, deleted or carry an unparsable filter were
left behind and kept products in dynamic groups they no longer belong to. The
existing mappings of the given products are now loaded across all streams and the
ones that belong to a stream which was not evaluated are removed.

vienthuong committed:
fix(core): keep display groups of all variants in sync

The skip condition read the display group of a single variant, because
`fetchAllKeyValue()` collapses all rows of a parent into one entry. Variants of the
same parent can hold diverging values, so whenever the sampled variant was already
up to date the update was skipped for all of its siblings and a variant without a
display group stayed hidden from listings. The number of outdated variants is now
counted in SQL.

vienthuong committed:
chore(core): drop ignored PHPStan error in ProductStreamUpdater

Danger rejects touched files that still carry baselined PHPStan errors. Replace the
remaining `empty()` with an explicit array check and remove the baseline entry.

vienthuong committed:
test(core): cover the search keyword diffing

Add unit coverage for the decision which keywords are deleted and inserted, so the
new diff based write path is asserted instead of only its integration behaviour.

`ProductCategoryDenormalizer` is a pure SQL denormalizer that is covered by its
dedicated integration test, mark it accordingly like `VariantListingUpdater`.

vienthuong committed:
fix(core): retry the keyword delete on deadlocks

The removed bulk delete ran inside `RetryableQuery::retryable()`. The targeted delete
that replaced it was a plain statement, so a deadlock during concurrent indexing was
no longer retried. Restore the wrapper.

vienthuong committed:
fix(core): keep repairing cheapest price accessors of unchanged containers

Skipping the whole product when its serialized container did not change also skipped
the per variant accessor check and the `ProductIndexerEvent` that refreshes SEO urls,
so a variant whose accessor got out of sync was never repaired again. Only the
`cheapest_price` write is skipped now, the accessor loop has its own guard per variant.

vienthuong committed:
fix(core): write stream mapping changes in one transaction

The deletes ran per stream inside the loop while the inserts were flushed afterwards,
so readers could observe a product that was already removed from a stream but not yet
re-inserted into another. Collect the deletes and execute them together with the insert
queue in a single retryable transaction.

vienthuong committed:
perf(core): delete search keywords of changed products in one statement

The keyword diff issued one delete per product, which is worse than the single bulk
delete it replaced whenever keywords actually change, e.g. during an import or a full
reindex. Compare the stored keywords and rankings to decide whether a product changed
at all, then rewrite the products that did with one batched delete. Unchanged products
are still skipped entirely, which is where the win comes from.

vienthuong committed:
docs: add release info for the diff based product indexing

Document that the product indexer only writes changed rows now, including the effect
on anything that observes row churn in the denormalized tables.

vienthuong committed:
fix(core): write keywords and dictionary in one transaction

The keyword rows and the dictionary rows they imply were committed separately, so an
interrupted run could leave a product indexed without its dictionary entries. No later
run repaired that, because the product's keywords then compare equal and the product is
skipped. Both writes and the delete now share one transaction.

vienthuong committed:
perf(core): delete category tree rows of a batch in one statement

The targeted delete ran inside the per product loop, each in its own retryable
transaction, so a batch opened one transaction per changed product where the bulk
delete it replaced used one for the whole batch. Collect the rows to remove and
delete them per chunk, keyed by `(product_id, category_id)` so the primary key is
still used. Same fix as the one applied to the search keyword delete.

vienthuong committed:
fix(core): repair the category tree column when it drifted from the rows

`product.category_tree` was only rewritten when the tree rows changed, so a column that
went stale while the rows stayed correct was never repaired and the product silently
dropped out of category listings that read it. The column is now carried along by the
mapping query and compared as a set, so it is written whenever it differs while runs
that change nothing still write nothing.

vienthuong committed:
fix(core): restore missing keyword dictionary rows

The dictionary is part of what a rebuild used to write, but it was not part of what the
diff compares: rows were only written for products whose keywords changed, so a
dictionary emptied out of band was never rebuilt. Migrations do exactly that, truncate
it and register the indexer, and `KeywordLoader` reads it to expand a search term, so
the search stayed degraded with no run able to repair it.

The analyzed keywords of every product, including the skipped ones, are now checked
against the dictionary and only the absent rows are written. That also drops the
previous unconditional REPLACE of rows that already exist.

vienthuong committed:
Merge branch 'trunk' into feat/optimize-product-indexing-to-avoid-unnecessary-table-refresh
